apps/aecore/lib/aecore/contract/tx/contract_call_tx.ex

```diff
@@ -0,0 +1,377 @@
+       end
+
+     # Insert the call into the state tree. This is mainly to remember what the
+     # return value was so that the caller can access it easily.
+     # Each block starts with an empty calls tree.
+     updated_calls_tree = CallStateTree.insert_call(calls, call)
+
+     {:ok, %{update_chain_state1 | accounts: accounts2, calls: updated_calls_tree}}
+   end
+
+   @spec preprocess_check(
+           Chainstate.accounts(),
+           tx_type_state(),
+           non_neg_integer(),
+           ContractCallTx.t(),
+           DataTx.t(),
+           Transaction.context()
+         ) :: :ok | {:error, String.t()}
+   def preprocess_check(
+         accounts,
+         _calls,
+         _block_height,
+         %ContractCallTx{amount: amount, gas: gas, gas_price: gas_price} = call_tx,
+         data_tx,
+         context
+       )
+       when is_non_neg_integer(gas_price) do
+     sender = DataTx.main_sender(data_tx)
+     fee = DataTx.fee(data_tx)
+
+     chain_state = Chain.chain_state()
+
+     required_amount = fee + gas * gas_price + amount
+
+     checks =
+       case context do
+         :transaction ->
+           [
+             fn -> check_validity(call_tx.call_stack == [], "Non empty call stack") end,
+             fn -> check_account_balance(accounts, sender, required_amount) end,
+             fn -> check_call(call_tx, chain_state) end
+           ]
+
+         :contract ->
+           [
+             fn -> check_call(call_tx, chain_state) end,
+             fn -> check_contract_balance(accounts, sender, amount) end
+           ]
+       end
+
+     case validate_fns(checks) do
+       :ok -> :ok
+       {:error, message} -> {:error, message}
+     end
+   end
+
+   @spec deduct_fee(
+           Chainstate.accounts(),
+           non_neg_integer(),
+           ContractCallTx.t(),
+           DataTx.t(),
+           non_neg_integer()
+         ) :: Chainstate.accounts()
+   def deduct_fee(accounts, block_height, _tx, data_tx, fee) do
+     DataTx.standard_deduct_fee(accounts, block_height, data_tx, fee)
+   end
+
+   def encode_to_list(%ContractCallTx{} = tx, %DataTx{} = datatx) do
+     [sender] = datatx.senders
+
+     [
+       :binary.encode_unsigned(@version),
+       Identifier.encode_to_binary(sender),
+       :binary.encode_unsigned(datatx.nonce),
+       Identifier.encode_to_binary(tx.contract),
+       :binary.encode_unsigned(tx.vm_version),
+       :binary.encode_unsigned(datatx.fee),
+       :binary.encode_unsigned(datatx.ttl),
+       :binary.encode_unsigned(tx.amount),
+       :binary.encode_unsigned(tx.gas),
+       :binary.encode_unsigned(tx.gas_price),
+       tx.call_data
+     ]
+   end
+
+   def decode_from_list(@version, [
+         encoded_sender,
+         nonce,
+         encoded_contract,
+         vm_version,
+         fee,
+         ttl,
+         amount,
+         gas,
+         gas_price,
+         call_data
+       ]) do
+     payload = %ContractCallTx{
+       contract: encoded_contract,
```

Comment:
You should decode the Identifier.

Reply:
This might not be in scope of this PR, but such errors should be caught by tests.